Web/Edubase.Data/Repositories/NotificationBannerRepository.cs

```diff
@@ -40,35 +48,81 @@ public async Task CreateAsync(params NotificationBanner[] entities)
+        public IEnumerable<NotificationBanner> GetTopTwoCurrentNotificationBanners()
         {
-            var query = Table.CreateQuery<NotificationBanner>().Where(x => x.PartitionKey == partitionKey.ToString()).AsQueryable();
-            if (visibleOnly)
+            var result = _cache[TopTwoNotificationBannersCacheKey] as IEnumerable<NotificationBanner>;
+            return result ?? GetTopTwoNotificationBannersFromTableAndUpdateCache();
+        }
+
+        private IEnumerable<NotificationBanner> GetTopTwoNotificationBannersFromTableAndUpdateCache()
+        {
+            var top2CurrentQuery =
+                Table.CreateQuery<NotificationBanner>()
+                    .Where(x => x.PartitionKey == eNotificationBannerPartition.Current.ToString()
+                                && x.Start <= DateTime.Now
+                                && x.End >= DateTime.Now)
+                    .Take(2).AsTableQuery();
+            var notificationBanners = Table.ExecuteQuery(top2CurrentQuery).ToList();
+
+            //the cache only adds values if they don't exist - the method name doesn't make this clear.
+            _cache.Add(TopTwoNotificationBannersCacheKey,
+                notificationBanners,
+                new DateTimeOffset(CalculateDateTimeOfCacheExpiration(notificationBanners)));
+
+            return notificationBanners;
+        }
+
+        private DateTime CalculateDateTimeOfCacheExpiration(IEnumerable<NotificationBanner> currentBanners)
+        {
+            if (!int.TryParse(ConfigurationManager.AppSettings["NotificationBannerCacheExpirationInSeconds"],
+                    out var maximumCacheExpirationInSeconds))
             {
-                query = query.Where(x => x.Start <= DateTime.Now && x.End >= DateTime.Now);
+                maximumCacheExpirationInSeconds = 300;
             }
-            query = query.Take(take);
-            var results = Table.ExecuteQuerySegmentedAsync(query.AsTableQuery(), skip).Result;
-            return new Page<NotificationBanner>(results, results.ContinuationToken);
+
+            var futureBannersQuery = Table.CreateQuery<NotificationBanner>()
+                .Where(x => x.PartitionKey == eNotificationBannerPartition.Current.ToString()
+                            && x.Start > DateTime.Now)
+                .AsTableQuery();
+            var futureBanners = Table.ExecuteQuery(futureBannersQuery).ToList();
+
+            var expirationTimes = (from banner in currentBanners
+                select banner.End).ToList();
+
+            expirationTimes = expirationTimes.Union((from banner in futureBanners
+                select banner.Start).ToList()).ToList();
+
+            expirationTimes.Add(DateTime.Now.AddSeconds(maximumCacheExpirationInSeconds));
+
+            return expirationTimes.Min();
```

Comment:
```
// Find the next time that the list of current banners changes (or cache TTL expires) -- either when:
// - a new banner is to appear, or
// - a current one is to expire, or
// - cache TTL expires (`NotificationBannerCacheExpirationInSeconds`)
```

Reply:
I have a new way of doing this that avoids the complication and only one trip to storage